python/ccf/ledger.py

```diff
@@ -149,6 +164,200 @@ def _byte_read_safe(file, num_of_bytes):
+        Depending on the tables that were part of the transaction, it does different things.
+        When transaction contains signature table, it starts the verification process and verifies that the root of merkle tree was signed by a node which was part of the network.
+        It also matches the root of the merkle tree that this class maintains with the one extracted from the ledger.
+        If any of the above checks fail, this method throws.
+        """
+        # Start with empty bytes array. CCF MerkleTree uses an empty array as the first leaf of it's merkle tree.
+        # Don't hash empty bytes array.
+        self.signature_ending = False
+        transaction_public_domain = transaction.get_public_domain()
+        tables = transaction_public_domain.get_tables()
+
+        # Add contributing nodes certs and update nodes network trust status for verification
+        if NODES_TABLE_NAME in tables:
+            node_table = tables[NODES_TABLE_NAME]
+            for nodeid, values in node_table.items():
+                # Add the nodes certificate
+                self.node_certificates[nodeid] = values[self.EXPECTED_NODE_CERT_INDEX]
+                # Update node trust status
+                self.node_activity_status[nodeid] = NodeStatus(
+                    values[self.EXPECTED_NODE_STATUS_INDEX]
+                )
+
+        # This is a merkle root/signature tx if the table exists
+        if SIGNATURE_TX_TABLE_NAME in tables:
+            self.signature_ending = True
+            self.signature_count += 1
+            signature_table = tables[SIGNATURE_TX_TABLE_NAME]
+
+            for nodeid, values in signature_table.items():
+                signing_node = int(
+                    values[self.EXPECTED_NODE_SIGNATURE_INDEX][
+                        self.EXPECTED_SIGNING_NODE_ID_INDEX
+                    ]
+                )
+                # Get binary representations for the cert, existing root, and signature
+                cert = b"".join(self.node_certificates[signing_node])
+                existing_root = b"".join(values[self.EXPECTED_ROOT_INDEX])
+                signature = values[self.EXPECTED_NODE_SIGNATURE_INDEX][
+                    self.EXPECTED_SIGNATURE_INDEX
+                ]
+
+                tx_info = TxBundleInfo(
+                    self.merkle,
+                    existing_root,
+                    cert,
+                    signature,
+                    self.node_activity_status,
+                    signing_node,
+                )
+
+                # validations for 1, 2 and 3
+                # throws if ledger validation failed.
+                if self._verify_tx_set(tx_info):
+                    continue
+                else:
+                    raise ValueError("Ledger validation failed")
+
+        # Checks complete, add this transaction to tree
+        self.merkle.add_leaf(transaction.get_public_tx())
```

Comment:
Could you rename `get_public_tx()` as it returns the entire serialised transaction as opposed to just the public doamin?